### DISCUSSION TITLE
Higher-level API development

### DISCUSSION THREAD

**Discussion by KickItLikeShika** [pytorch/ignite#1641] [Show and tell]
I have experimented a lot of different ideas for this Higher-level API, but at the end i went with this one. It's obviously not done yet so i got some questions:
1- What do you think so far? is there anything i should do? anything i should replace? better implementation? better idea? better way to do this?
2- I got an error provided in the github gist below, i need help, or a better way to implement it

note:
- The reason why i chose this way to implement the trainer.fit() just because i found it so cool! like the one provided in sklearn, define the trainer, then fit it, then predict. And all of the this work fine btw.

File on Github Gist:
[Trainer Implementation](https://gist.github.com/KickItLikeShika/e3f31a19ae476c923d2e2416db42c070)

**Comment by vfdev-5** [comment]
@KickItLikeShika thanks for exploring this question! I think discussing that we could find something interesting and helpful.
Maybe, a generic though about that: please, keep in mind that adopted approach have to be "generic" = user could handle single/multiple models/optimizers, data sources etc.

I think, it would be very helpful to take a look and analyze existing solutions referenced in the issue and have a clear idea of what is already done and the pros/cons of those approaches.

What do you think ?

**Reply by KickItLikeShika** [reply]
@vfdev-5 Sure i will keep exploring and updating, but about the idea behind the fit method/validation method and the other methods what do you think? because it's obviously not similar to create_supervised_trainer. The fit method returns nothing, and i'm trying to make the validation method just return the metrics. And is there any other methods should i work on? or any ideas you got on your mind?

**Reply by vfdev-5** [reply]
Yes, a Model API with fit/validate/predict is something interesting to have. True that this API takes more responsibilities vs basic `create_supervised_trainer`. Return type is a detail, I think at this stage. Maybe, another detail to have in mind also is to expose publicly the engine such that user could add really custom features to the training loop.

I'd say by priority: 
- analysis of other solutions
- implement more complicated training schemas (GAN's, Semi-supervised learning, ...) with experimental API
- how to handle out-of-the-box distributed training, automatic mixed precision option, etc in a reasonable way...

**Reply by KickItLikeShika** [reply]
I really appreciate your help! Thank you so much!
About the error in `validate` method, do you have any idea? or if you were implementing this, how would you do it? like were you going to return the metrics value? or you have a better idea?

**Reply by vfdev-5** [reply]
Currently, in the library we return all the time `engine.state` from any run. First idea would be to return it as well in all those fit/validate/predict methods such that if user registers something more than metrics then it is still available...

**Reply by KickItLikeShika** [reply]
Thank you sir! I will keep working

**Reply by KickItLikeShika** [reply]
Hey @vfdev-5, now the prototype works as expected with the basic requirements (fit/validate/predict). The first thing you said "It has to be more generic". I want to hear your ideas to make it more generc! i will really appreciate you well. Also i want to hear from @sdesrozis.
The File on Github Gist:
[Trainer Implementation](https://gist.github.com/KickItLikeShika/0d88071c17192300b9ca91ac7315b555)

**Reply by vfdev-5** [reply]
Hey @KickItLikeShika it is up to GSoC project to propose few ideas on how to make it more generic :)
But again, please read this comment https://github.com/pytorch/ignite/discussions/1641#discussioncomment-368713 and 
start with "analysis of other solutions". This is **very very important step** to accomplish before starting coding your own solution. As soon as you better understand others solution, better you could asses your own proposition.
Once, you are done with the first step, try the second step "implement more complicated training schemas (GAN's, Semi-supervised learning, ...) with experimental API".

**Reply by KickItLikeShika** [reply]
@vfdev-5 Okay i got it, Thank you!

**Reply by vfdev-5** [reply]
Yes, please update this thread once you are done with the first step and could share a small doc on that.

**Reply by sdesrozis** [reply]
@KickItLikeShika Thank you for your solution. I totally agree @vfdev-5, we must first have a good overview of other solutions before thinking about our implementation. I am convinced that implementation is not the difficult part of this work. It is the analysis of the state of the art and the perspective on complex cases.

**Reply by KickItLikeShika** [reply]
Hey @vfdev-5, @sdesrozis, here is a document i have made that shows the pros and cons for each solution that i have walk through, if you got any other interesting solution please let me know and i will immediately walk through and analyze it then update the document, i tried to keep the doc as short as possible.
[Analysis of Higher-Level API Solution ](https://docs.google.com/document/d/1GkK5koa5D6iYbJd3Rwa2y_sKUm2jxBDji4hhBr1oHew/edit?usp=sharing)

**Reply by sdesrozis** [reply]
Thank you for this work. Having an exhaustive comparison of apis is a good idea. We must be aware to limit subjective opinion even if it is not completely possible. Provide sample codes would be very interesting. 

Maybe we could first list relevant apis ?

**Reply by KickItLikeShika** [reply]
Sure, tomorrow a bunch of sample codes will be ready and posted here. Can i join slack btw?

**Reply by vfdev-5** [reply]
We are using pytorch slack which we are not admin, so if you are already on it you can find us on #pytorch-ignite channel. Otherwise, I'm using discord to communicate with folks (vfdev-5#4458). On the other hand, it is also nice to use github discussions such that things could help the community...

**Reply by KickItLikeShika** [reply]
@vfdev-5 @sdesrozis 
I did as @sdesrozis asked. I have provided some sample codes and made APIs similar to MONAI, Transfer-NLP, ARGUS, trainers. I really tried to make it much smaller and made some function raise `NotImplementedError` just to worry more about the idea. I don't provide a lot features just for the simplicity for now, and sure used PyTorch-Ignite `engine`. I hope it's clear enough for you to understand the idea. I see you updated [GSoC 2021 Development of a Higher-level API](https://github.com/pytorch/ignite/wiki/GSoC-2021-project#development-of-a-higher-level-api) and it's more clear now and i hope you keep updating it and provide more details. And what do you think is gonna be more interesting? generic API like Argus/MONAI or task specific API like Transfer-NLP? Both seem really interesting. I wait for you reply to start building the API.

[Implementation Similar To MONAI
](https://gist.github.com/KickItLikeShika/72174e95aa471a7aaa95ba9ea6cb9acc)

[Implementation Similar to Transfer-NLP
](https://gist.github.com/KickItLikeShika/08c67757b20c5bbac412e55d44f4483f)

[Implementation Similar to Argus](https://gist.github.com/KickItLikeShika/c0c23906807394714e1a98ce5b429b37)

**Reply by sdesrozis** [reply]
@KickItLikeShika Thank you ! I take look at your samples as soon as possible.
